---
title: "02_compara"
author: "zagor"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    self_contained: yes
    fig_width: 9
    fig_height: 8
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 5
    number_sections: yes
    theme: flatly
    highlight: tango
  pdf_document:
    toc: yes
    toc_depth: '2'
  word_document:
    toc: yes
    toc_depth: '2'
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = c('png', 'pdf'),  # this embeds pdf and crates scrolable blocks if pdf is first
                      fig.path = "../reports/compara/", # save each figure here
                      # dev = c('png'), 
                      fig.align = 'center', 
                      dpi = 300,
                      fig.height = 8, 
                      fig.width = 9 ,
                      warning = FALSE, message = FALSE
                      )
```



```{r,  echo=TRUE, warning=FALSE, message=FALSE}

rm(list = ls(all = TRUE))
gc()


library(magrittr)
library(ggplot2)
library(data.table)

`%nin%` = Negate(`%in%`)


setwd(dirname(rstudioapi::getActiveDocumentContext()$path))




```


![](../figs/tree_example1.png)

<https://www.ensembl.org/info/genome/compara/homology_types.html>

<https://www.ensembl.org/info/website/tutorials/compara.html>

* Homologues are separated into Orthologues (across different species) and Paralogues (within a species).
* Whole genome alignments are determined as part of the Ensembl release cycle, incorporating new species and new assemblies.

<https://www.ensembl.org/info/genome/compara/homology_method.html>

<https://www.ensembl.org/info/genome/compara/Ortholog_qc_manual.html>

* The whole genome alignment score calculates the coverage of the alignment over the orthologue pair.
* In every orthology inference between two species, the set of orthologue pairs between the species is considered for high-confidence annotation. The GOC and WGA scores among others are used to determine which orthologies to tag as being high-confidence.
*  The primary filter used to select the high-confidence set of orthologue pairs, consists of three thresholds for respectively the GOC score, the WGA coverage and the percentage identity of the orthologue pairs. The thresholds we use depend on the most recent common ancestor of the species pair, according to the table below. The primary filter is used if there are scores of either type (GOC or WGA) in the set of orthologues between the two species. 
* The orthology between two genes will be classified as high confidence if the alignment between the two genes satisfies the percentage identity requirement and either GOC or WGA scores meet their thresholds.
* If GOC or WGA scores are not available or if no threshold is used for a set of orthologues, the pipeline will use fallback criteria, that is the prediction will be classified as high confidence if the percentage identity meets the threshold and the prediction is [tree compliant](https://www.ensembl.org/info/genome/compara/homology_types.html#betweenspeciesparalogues). 



```{r, clean}

clean_text = function(x) {
  if (is.character(x)) {
    # Remove all whitespace and quotes
    cleaned <- gsub("[\\s\"']", "", x, perl = TRUE)
    # Replace "NULL" with NA
    ifelse(cleaned == "NULL", NA, cleaned)
  } else {
    x
  }
}

```



# plant selection

```{r, sepcies}

fp = file.path('..', 'input', 'Compara')

species = c('arabidopsis_thaliana', 
            'malus_domestica_golden',
            'prunus_avium', 
            'prunus_dulcis', 
            'prunus_persica',
            'solanum_lycopersicum',
            'solanum_tuberosum',
            'vitis_vinifera')

```

# stu exception

```{r, stu_read}

# download and read Excel files
read_excel_from_url <- function(url, start_row = 1, columns = NULL) {
  temp_file = tempfile(fileext = ".xlsx")
  utils::download.file(url, destfile = temp_file, mode = "wb")
  data = openxlsx::read.xlsx(temp_file, startRow = start_row)
  if (!is.null(columns)) data = data[, columns]
  return(data)
}

# DMT2DMG in unitato
dmt_url = "https://raw.githubusercontent.com/NIB-SI/unitato/main/output/Phureja_v4-v6.1_annotations.xlsx"
dmt_cols = c("geneID", "v4_PGSC_DMT")
dmt_data = read_excel_from_url(dmt_url, start_row = 3, columns = dmt_cols)
dmt_data = dmt_data[!is.na(dmt_data$v4_PGSC_DMT), ]
data.table::setDT(dmt_data)
dmt_data.unitato = data.table::copy(dmt_data)

# converter table
zip_url = "https://github.com/NIB-SI/unitato/raw/main/input/Solanum_tuberosum_PGSC_DM_v3.4_converterWithDescriptions.txt.zip"
zip_file = tempfile(fileext = ".zip")
temp_dir = tempdir()
utils::download.file(zip_url, zip_file, mode = "wb")
utils::unzip(zip_file, exdir = temp_dir)
converter_file = file.path(temp_dir, "Solanum_tuberosum_PGSC_DM_v3.4_converterWithDescriptions.txt")
converter_data = data.table::fread(converter_file, sep = "\t", header = FALSE, fill = TRUE, quote = "")
converter_data = converter_data[, .(V1, V2)]
data.table::setnames(converter_data, names(dmt_data))
dmt_data = unique(rbind(dmt_data, converter_data))
data.table::setnames(dmt_data, c("sourceID", "Ortholog"))

# Load and clean translation table
trans_url = "https://raw.githubusercontent.com/NIB-SI/unitato/main/output/Phureja_v4-v6.1_translations.xlsx"
trans_cols = c("v4_geneID", "v6.1_geneID")
trans_data = read_excel_from_url(trans_url, start_row = 2, columns = trans_cols)
colnames(trans_data) = c("Ortholog", "sourceID")
trans_data = trans_data[grep("PGSC", trans_data$Ortholog), ]

# unitato table ++
dmt_data = dmt_data[sourceID %nin% trans_data$Ortholog]
dmt_data[, Ortholog := sourceID]
unitato = rbind(trans_data, dmt_data)
unitato = unitato[]
data.table::setDT(unitato)
unitato <- unitato[, .(sourceID, Ortholog)]

lost = setdiff(grep("PGSC", unitato$sourceID, value = TRUE), dmt_data.unitato$geneID)
table(grep("PGSC", unitato$sourceID, value = TRUE) %in% dmt_data.unitato$geneID)
unitato = unitato[unitato$Ortholog %nin% lost, ]


```



# params

_Denote:_ check (input fasta nad output orthofider) files to decide on regex

ppe and pdul exceptions - mapping by protID

```{r, params}

params_list = list(
  
  list(to_plant = 'arabidopsis_thaliana', 
       stp = 'ath' # standardised plant abbreviation 
       ),
    
  list(to_plant = 'solanum_tuberosum', 
       stp = 'stu', # standardised plant abbreviation 
       ref_genome = "UniTato_pep_oneliner",
       compara_genome = "Compara_Solanum_tuberosum.SolTub_3.0.pep.all",
       pdir = "stu_potato",
       from_sub_pattern = NULL,
       swap_sub_pattern = "",
       to_sub_pattern = NULL
       ),
  list(to_plant = 'solanum_lycopersicum', 
       stp = 'sly', # standardised plant abbreviation 
       ref_genome = "Phytozome_Slycopersicum_796_ITAG5.0.protein_primaryTranscriptOnly",
       compara_genome = "Compara_Solanum_lycopersicum.SL3.0.pep.all",
       pdir = "sly_tomato",
       from_sub_pattern = "^(([^.]*\\.){1}[^.]*)\\..*$", # keep everything before the second dot
       swap_sub_pattern = "\\1",
       to_sub_pattern = "(\\.[0-9]+)+$"), 
  list(to_plant = 'vitis_vinifera', 
       stp = 'vvi', # standardised plant abbreviation 
       ref_genome = "Grapedia_5.1_on_T2T_ref_main_proteins",
       compara_genome = "Compara_Vitis_vinifera.PN40024.v4.pep.all",
       pdir = "vvi_grapevine",
       from_sub_pattern = "_.*",
       swap_sub_pattern = "",
       to_sub_pattern = "(_[^_]*){2}$"), # after second _
    
  list(to_plant = 'malus_domestica_golden', 
       stp = 'mdo', # standardised plant abbreviation 
       ref_genome = "g.Honeycrisp_HAP1_braker1+2_combined_fullSupport_longname_filtered.pep",
       compara_genome = "Compara_Malus_domestica_golden.ASM211411v1.pep.all",
       pdir = "mdo_apple",
       from_sub_pattern = "^(?:mRNA_|CDS_)?(MD\\d+G\\d+).*",
       swap_sub_pattern = "\\1",
       to_sub_pattern = "\\.[^.]*$"),
  
  list(to_plant = 'prunus_persica', 
       stp = 'ppe', # standardised plant abbreviation 
       ref_genome = "PLAZA_proteome.selected_transcript.ppe",
       compara_genome = "Compara_Prunus_persica.Prunus_persica_NCBIv2.pep.all", # by protein
       pdir = "ppe_peach",
       from_sub_pattern = NULL,
       swap_sub_pattern = "",
       to_sub_pattern = "\\.[^.]*$"), # everythin after the last dot
  
    list(to_plant = 'prunus_dulcis', 
       stp = 'pdul', # standardised plant abbreviation 
       ref_genome = "Texas_F1_protein",
       compara_genome = "Compara_Prunus_dulcis.ALMONDv2.pep.all",
       pdir = "pdul_almond",
       from_sub_pattern = NULL, # by protein
       swap_sub_pattern = "",
       to_sub_pattern = "\\.[^.]*$"),
  
    list(to_plant = 'prunus_avium', 
       stp = 'pavi', # standardised plant abbreviation 
       ref_genome = "Prunus_avium_Tieton.proteins",
       compara_genome = "Compara_Prunus_avium.PAV_r1.0.pep.all",
       pdir = "pavi_wildCherry",
       from_sub_pattern = "^(([^_]+_){2}[^_]+).*", 
       swap_sub_pattern = "\\1",
       to_sub_pattern = "^[^ ]+\\s+") 
  
)


pattern_mapP = purrr::map_chr(params_list, "stp") %>%
  setNames(purrr::map_chr(params_list, "to_plant"))


```

# orthofinder - within

```{r, orthofinder}


# Function to process one parameter set
process_param <- function(param) {
  
  stp = param$stp
  print(stp)
  ref_genome = paste0(param$ref_genome, '.tsv')
  compara_genome = param$compara_genome
  to_sub_pattern = param$to_sub_pattern
  swap_sub_pattern = param$swap_sub_pattern
  from_sub_pattern = param$from_sub_pattern
  pdir = param$pdir
  
  if (stp == "stu") { # potato exception
    result = unitato
  } else {
    

    files = list.files(file.path("..", "input", "OrthoFinder", pdir))
    fn = grep(ref_genome, files, value = TRUE, fixed = TRUE)
      
    compara = data.table::fread(file.path("..", "input", "OrthoFinder", pdir, fn), header = TRUE, sep = "\t", stringsAsFactors = FALSE)
    
    data.table::setnames(compara, old = colnames(compara)[3], new = "reference")
      
    if (!grepl(compara_genome, ref_genome)) {
      compara = compara[Species == compara_genome]
    } else {
      compara[, Orthologs := reference]
    }
      
    compara[, source_list := stringr::str_split(reference, pattern = ",\\s*")]
    compara[, orthologs_list := stringr::str_split(Orthologs, pattern = ",\\s*")]
      
    result = compara[, {
      pairs = data.table::CJ(source_list[[1]], orthologs_list[[1]], sorted = FALSE)
      data.table::setnames(pairs, c("sourceID", "Ortholog"))
      pairs
    }, by = seq_len(nrow(compara))]
    result[, seq_len := NULL]
    
    if (!is.null(to_sub_pattern)) {
      result[, sourceID := gsub(to_sub_pattern, "", sourceID)]
    }
    if (!is.null(from_sub_pattern)) {
      result[, Ortholog := gsub(from_sub_pattern, swap_sub_pattern, Ortholog)]
    } 
    
  }
    

  result$to_plant = stp
  unique(result)
  
}

# Use lapply to process list, ignore ath
reference_list = lapply(params_list[-1], process_param)
reference = data.table::rbindlist(reference_list)
rm(reference_list)

```


# merge

```{r, merge}

fl = list.files(fp, pattern = '\\.zip')

comparaPlants = NULL

for (i in fl) {
  
  # print(i)
  
  df = data.table::fread(file.path(fp, i))

  df = as.data.frame(lapply(df, clean_text), stringsAsFactors = FALSE)

  # print(table(df$homology_type))
  # print(apply(df, 2, function(x) sum(is.na(x))))
  
  df$source = sub('_.*', '', i)
  sort(unique(df$homology_species))
  
  df = df[df$homology_species %in% species, ]
  # table(df$species, df$homology_species)
  
  comparaPlants = rbind(comparaPlants, df)
  
}



table(comparaPlants$species, comparaPlants$homology_species)
table(comparaPlants$homology_type, comparaPlants$source)
sum(duplicated(comparaPlants$homology_id))

ggplot(comparaPlants, aes(x = homology_identity)) +
  geom_histogram(binwidth = 1, fill = "steelblue", color = "black") +
  facet_wrap(~ source, ncol = 4) +
  theme_minimal() +
  labs(
    title = "Distribution of Homology Identity per source table",
    x = "Homology Identity",
    y = "Count"
  )

ggplot(comparaPlants, aes(x = identity)) +
  geom_histogram(binwidth = 1, fill = "steelblue", color = "black") +
  facet_wrap(~ source, ncol = 4) +
  theme_minimal() +
  labs(
    title = "Distribution of Identity per source table",
    x = "Identity",
    y = "Count"
  )



table(comparaPlants$species, comparaPlants$homology_species)

ind = intersect(grep('prunus|arabidopsis', comparaPlants$species), grep('prunus|arabidopsis', comparaPlants$homology_species))
table(comparaPlants$species[ind], comparaPlants$homology_species[ind])

ind = intersect(grep('solanum|arabidopsis', comparaPlants$species), grep('solanum|arabidopsis', comparaPlants$homology_species))
table(comparaPlants$species[ind], comparaPlants$homology_species[ind])

```


# high confidence

```{r, hc}

ind = which(comparaPlants$is_high_confidence == 1)
comparaPlants.hc = comparaPlants[ind, ]
rm(comparaPlants)
gc()
table(comparaPlants.hc$species, comparaPlants.hc$homology_species)

ind = intersect(grep('prunus|arabidopsis', comparaPlants.hc$species), grep('prunus|arabidopsis', comparaPlants.hc$homology_species))
table(comparaPlants.hc$species[ind], comparaPlants.hc$homology_species[ind])

ind = intersect(grep('solanum|arabidopsis', comparaPlants.hc$species), grep('solanum|arabidopsis', comparaPlants.hc$homology_species))
table(comparaPlants.hc$species[ind], comparaPlants.hc$homology_species[ind])


ggplot(comparaPlants.hc, aes(x = homology_identity)) +
  geom_histogram(binwidth = 1, fill = "steelblue", color = "black") +
  facet_wrap(~ source, ncol = 4) +
  theme_minimal() +
  labs(
    title = "Distribution of Homology Identity per source table",
    x = "Homology Identity",
    y = "Count"
  )



# file_path = '../intermediate/comparaPlants.high_confidence.txt'
# data.table::fwrite(comparaPlants.hc, file_path, sep = '\t')
# root_dir = dirname(file_path)
# file_name = basename(file_path)
# zip::zip(zipfile = paste0(file_path, ".zip"), 
#     files = file_name, 
#     root = root_dir)
# file.remove(file_path)


```

# with ath pairs

```{r, ath-pairs}

comparaPlants.filtered = comparaPlants.hc[
  grepl("arabidopsis", comparaPlants.hc$species, ignore.case = TRUE) |
    grepl("arabidopsis", comparaPlants.hc$homology_species, ignore.case = TRUE), ]
rm(comparaPlants.hc)
gc()
table(comparaPlants.filtered$species, comparaPlants.filtered$homology_species)


# file_path = '../intermediate/comparaPlants_high_confidence_ath_pairs.txt'
# data.table::fwrite(comparaPlants.filtered, file_path, sep = '\t')
# root_dir = dirname(file_path)
# file_name = basename(file_path)
# zip::zip(zipfile = paste0(file_path, ".zip"), 
#     files = file_name, 
#     root = root_dir)
# file.remove(file_path)

```

# ath-plant pairs

```{r, ath-pairs-reordered}

cols_to_keep = c("gene_stable_id", "protein_stable_id", "species", 
                 "homology_gene_stable_id", "homology_protein_stable_id", "homology_species",
                 "homology_identity")

comparaPlants.filtered = comparaPlants.filtered[, cols_to_keep, drop = FALSE]

ind1 = grep("arabidopsis", comparaPlants.filtered$homology_species, ignore.case = TRUE)
ind2 = grep("arabidopsis", comparaPlants.filtered$species, ignore.case = TRUE)


df_sub2 = comparaPlants.filtered[ind2, ]
cols_first = c(
  "gene_stable_id", "protein_stable_id", "species", 
  "homology_gene_stable_id", "homology_protein_stable_id", "homology_species",
  "homology_identity"
)
df_sub2 = df_sub2[, cols_first]


df_sub1 = comparaPlants.filtered[ind1, ]
cols_first = c(
  "homology_gene_stable_id", "homology_protein_stable_id", "homology_species",
  "gene_stable_id", "protein_stable_id", "species", 
  "homology_identity"
)
df_sub1 = df_sub1[, cols_first]

colnames(df_sub1) = colnames(df_sub2)
rm(comparaPlants.filtered)
gc()


df = rbind(df_sub2, df_sub1)
setDT(df)
df[, protein_stable_id := NULL]
# rm(df_sub1)
# rm(df_sub2)
df[, species := pattern_mapP[species]]
df[, homology_species := pattern_mapP[homology_species]]

ggplot(df, aes(x = homology_identity)) +
  geom_histogram(binwidth = 1, fill = "steelblue", color = "black") +
  facet_wrap(~ homology_species, ncol = 4) +
  theme_minimal() +
  labs(
    title = "Distribution of hc Homology Identity to Arabidopsis",
    x = "Homology Identity",
    y = "Count"
  )

df[, homology_identity := NULL]
df[homology_species %in% c("ppe", "pdul"), homology_gene_stable_id := homology_protein_stable_id]
df[, homology_protein_stable_id := NULL]

colnames(df) = c("from",
                 "from_plant",
                 "to",
                 "to_plant")
df$source = 'compara'

```

# translate to ref geneID

```{r, translate}

df = reference[df, on = .(Ortholog = to, to_plant), 
                .(from, from_plant, to = sourceID, to_plant, source)]
df = unique(df)
df[is.na(to), .N, by = to_plant]
df = df[!is.na(df$to), ]
table(df$from_plant, df$to_plant)

df %>%
  dplyr::group_by(to_plant) %>%
  dplyr::slice_head(n = 2) %>%
  dplyr::bind_rows(df %>% dplyr::group_by(to_plant) %>% dplyr::slice_tail(n = 2)) %>%
  dplyr::arrange(to_plant) %>%
  dplyr::ungroup() -> first_last_two_per_plant
print(first_last_two_per_plant, n = nrow(first_last_two_per_plant))

```


# write and zip

```{r, zip}

file_path = '../intermediate/comparaPlants-hc_cleaned.txt'
data.table::fwrite(df, file_path, sep = '\t')
root_dir = dirname(file_path)
file_name = basename(file_path)
zip::zip(zipfile = paste0(file_path, ".zip"), 
    files = file_name, 
    root = root_dir)
file.remove(file_path)

```


# sessionInfo

```{r,  echo=TRUE, warning=FALSE, message=FALSE}

sessionInfo()

```

