---
title: "06_OrthoDB"
author: "zagor"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    self_contained: yes
    fig_width: 9
    fig_height: 8
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 5
    number_sections: yes
    theme: flatly
    highlight: tango
  pdf_document:
    toc: yes
    toc_depth: '2'
  word_document:
    toc: yes
    toc_depth: '2'
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = c('png', 'pdf'),  # this embeds pdf and crates scrolable blocks if pdf is first
                      fig.path = "../reports/OrthoDB/", # save each figure here
                      # dev = c('png'), 
                      fig.align = 'center', 
                      dpi = 300,
                      fig.height = 8, 
                      fig.width = 9 ,
                      warning = FALSE, message = FALSE
                      )
```



```{r,  echo=TRUE, warning=FALSE, message=FALSE}

rm(list = ls(all = TRUE))
gc()


library(magrittr)
library(data.table)

setwd(dirname(rstudioapi::getActiveDocumentContext()$path))




```


| OrthoDBspeciesID | speciesAbbr |
|------------------|-------------|
| 3750_0           | mdo         |
| 36596_0          | parm        |
| 42229_0          | pavi        |
| 3755_0           | pdul        |
| 3760_0           | ppe         |
| 4081_0           | sly         |
| 4113_1           | stu         |
| 4113_0           | stu         |
| 29760_1          | vvi         |
| 29760_0          | vvi         |


# OrthoDB raw

```{r, orthoDB_read}

fp = file.path('..', 'input', 'OrthoDB')
fn = 'ortho_pairs_eudicotyledons_top.tsv.zip'

df = data.table::fread(file.path(fp, fn))



```

_Deote:_ use gene IDs instead of protein/transcript IDs


# ath

```{r, ath_read}

fp = file.path('..', 'input', 'OrthoFinder', 'ath_arabidopsis')
fn = 'OrthoDB_Arabidopsis_thaliana_C24.tsv'

ath = data.table::fread(file.path(fp, fn))
columns = c("OrthoDB_Arabidopsis_thaliana_C24", "Orthologs")
ath = ath[, ..columns]


ath[, OrthoDB_list := stringr::str_split(OrthoDB_Arabidopsis_thaliana_C24, pattern = ",\\s*")]
ath[, Orthologs_list := stringr::str_split(Orthologs, pattern = ",\\s*")]
result = ath[, {
  # Cartesian join of OrthoDB_list and Orthologs_list for this row
  pairs = CJ(OrthoDB_list[[1]], Orthologs_list[[1]], sorted = FALSE)
  setnames(pairs, c("OrthoDB_ID", "Ortholog"))
  pairs
}, by = seq_len(nrow(ath))]
ath = result[, seq_len := NULL]
ath$Ortholog = sub("\\.[0-9]+.*$", "", ath$Ortholog)
ath = ath[!duplicated(ath), ]

colnames(ath)[2] = 'TAIR'
head(ath)


```

# stu exception

```{r, stu_read}

url = "https://raw.githubusercontent.com/NIB-SI/unitato/main/output/Phureja_v4-v6.1_annotations.xlsx"
temp_file = tempfile(fileext = ".xlsx")
download.file(url, destfile = temp_file, mode = "wb")
DMT2DMG = openxlsx::read.xlsx(temp_file, startRow  = 3)
# head(DMT2DMG)
columns = c('geneID', 'v4_PGSC_DMT')
DMT2DMG = DMT2DMG[, columns]
DMT2DMG = DMT2DMG[!is.na(DMT2DMG$v4_PGSC_DMT), ]
setDT(DMT2DMG)

```



# params

```{r, params}}


params_list = list(
  list(inDir = 'stu_potato', to_plant = 'stu', 
       fn = 'OrthoDB_Solanum_tuberosum_DM3.tsv', 
       refGenome = 'UniTato_pep_oneliner', 
       to_sub_pattern = "(\\.[0-9]+)+$"), # numbers after last dots marking gene ID: .1 or .1.1 + exception
  list(inDir = 'sly_tomato', to_plant = 'sly', 
       fn = 'OrthoDB_Solanum_lycopersicum_ITAG3.tsv', 
       refGenome = 'Phytozome_Slycopersicum_796_ITAG5.0.protein_primaryTranscriptOnly', 
       to_sub_pattern = "\\.[^.]*$"), # everythin after the last dot
  list(inDir = 'vvi_grapevine', to_plant = 'vvi', 
       fn = 'OrthoDB_Vitis_vinifera_PN40024_2009.tsv', 
       fnPlus = 'OrthoDB_Vitis_vinifera_PN40024_2021.tsv', 
       refGenome = 'Grapedia_5.1_on_T2T_ref_main_proteins', 
       to_sub_pattern = "(_[^_]*){2}$"), # remove everything second from the last '_'
    
  list(inDir = 'mdo_apple', to_plant = 'mdo', 
       fn = 'OrthoDB_Malus_domestica_GD.tsv', 
       refGenome = 'g.Honeycrisp_HAP1_braker1+2_combined_fullSupport_longname_filtered.pep', 
       to_sub_pattern = "\\.[^.]*$"), # everythin after the last dot
  list(inDir = 'ppe_peach', to_plant = 'ppe', 
       fn = 'OrthoDB_Prunus_persica_Lovell.tsv', 
       refGenome = 'PLAZA_proteome.selected_transcript.ppe',
       to_sub_pattern = "\\.[^.]*$"), # everythin after the last dot
  list(inDir = 'pdul_almond', to_plant = 'pdul',
       fn = 'OrthoDB_Prunus_dulcis_CNAG.tsv', 
       refGenome = 'Texas_F1_protein', 
       to_sub_pattern = "\\.[^.]*$"), # everythin after the last dot
  list(inDir = 'pavi_wildCherry', to_plant = 'pavi', 
       fn = 'OrthoDB_Prunus_avium_Satonishiki.tsv', 
       refGenome = 'Prunus_avium_Tieton.proteins', 
       to_sub_pattern = "-[^-]*$"), # T1, T2, T3 in .bed file have the same coord
  list(inDir = 'parm_apricot', to_plant = 'parm', 
       fn = 'OrthoDB_Prunus_armeniaca_Orared.tsv', 
       refGenome = 'Prunus_armeniaca_Marouch_n14_peptide', 
       to_sub_pattern = "\\.[^.]+\\.[^.]+$") # .t1.p1_1
)


```

# OrthoDB and OrthoFinder - within

```{r, f-ction}



process_OrthoDB_OrthoFinder <- function(inDir, fn, fnPlus, to_plant, refGenome, to_sub_pattern) {
  
  ## Orthoinder within
  # Read OrthoFinder result file
  fp = file.path('..', 'input', 'OrthoFinder', inDir)
  subset = data.table::fread(file.path(fp, fn))

  # Rename third column to OrthoDB (assuming expected format)
  colnames(subset)[3] = 'OrthoDB'
  
  
  if (!is.null(fnPlus)) {
    subsetPlus = data.table::fread(file.path(fp, fnPlus))
    colnames(subsetPlus)[3] = 'OrthoDB'
    subset = rbind(subset, subsetPlus)
  }
  
  

  # Filter for reference genome proteome
  subset = subset[subset$Species == refGenome, ]
  
  # Select relevant columns (3 and 4)
  colnames = c('OrthoDB', "Orthologs")
  # subset = subset[, c(3:4), with = FALSE]
  subset = subset[, ..colnames]
  
  # Split OrthoDB and Orthologs strings into lists
  subset[, OrthoDB_list := stringr::str_split(OrthoDB, pattern = ",\\s*")]
  subset[, Orthologs_list := stringr::str_split(Orthologs, pattern = ",\\s*")]
  
  # Cartesian join of OrthoDB_list and Orthologs_list for each row
  result = subset[, {
    pairs = CJ(OrthoDB_list[[1]], Orthologs_list[[1]], sorted = FALSE)
    setnames(pairs, c("OrthoDB_ID", "Ortholog"))
    pairs
  }, by = seq_len(nrow(subset))]
  
  subset = result[, seq_len := NULL]
  
  # Remove duplicate pairs
  subset = subset[!duplicated(subset)]
  
  df.subset = df[df$speciesAbbr == to_plant, ]
  
  # Replace : with _ in gene_IDs for consistent matching; original OrthoDB uses ":"
  df.subset$gene_ID1 = sub(':', '_', df.subset$gene_ID1)
  df.subset$gene_ID2 = sub(':', '_', df.subset$gene_ID2)
  
  # Merge with df.subset and subset to link IDs and orthologs
  combo = merge(df.subset, ath, by.x = 'gene_ID1', by.y = 'OrthoDB_ID', all.x = TRUE)
  combo = merge(combo, subset, by.x = 'gene_ID2', by.y = 'OrthoDB_ID', all.x = TRUE)
  # save if needed

  ind = grep('TAIR$|Ortholog$', colnames(combo))
  combo = combo[, ..ind]  
  # Clean NA entries
  ind = which(!is.na(combo[,1]) & !is.na(combo[,2]))
  combo = combo[ind, ]
  combo = combo[!duplicated(combo), ]
  
  # uniform colnames
  colnames(combo) = c('from', 'to')
  combo$from_plant = 'ath'
  combo$to_plant = to_plant
  combo$source = 'OrthoDB'
  
  # Reorder columns
  setcolorder(combo, c("from", "from_plant", "to", "to_plant", "source"))
  
  
  # Apply substitution on 'to' column if pattern provided
  if (!is.null(to_sub_pattern)) {
    
    combo$to = sub(to_sub_pattern, '', combo$to)
    
    if (to_plant == 'stu') {
      setkey(DMT2DMG, v4_PGSC_DMT)
      combo[DMT2DMG, on = .(to = v4_PGSC_DMT), to := geneID]
    }

    
  }
  
  combo = combo[!duplicated(combo), ]
  return(combo)
}



```

# translate pairs to rep genomes

```{r, f-ction_run}



# Initialize list
results = vector("list", length(params_list))

# Loop through parameter list
for (i in seq_along(params_list)) {
  
  print(params_list[[i]]$to_plant)
  p = params_list[[i]]
  
  results[[i]] = process_OrthoDB_OrthoFinder(
    
    inDir = p$inDir
    , fn = p$fn
    , fnPlus = p$fnPlus
    , to_plant = p$to_plant
    , refGenome = p$refGenome
    , to_sub_pattern = p$to_sub_pattern
    
  )
}


combo = rbindlist(results, use.names = TRUE, fill = TRUE)
combo = combo[!duplicated(combo), ]
table(combo$to_plant)


combo %>%
  dplyr::group_by(to_plant) %>%
  dplyr::slice_head(n = 2) %>%
  dplyr::bind_rows(combo %>% dplyr::group_by(source) %>% dplyr::slice_tail(n = 2)) %>%
  dplyr::arrange(source) %>%
  dplyr::ungroup() -> first_last_two_per_plant

print(first_last_two_per_plant, n = nrow(first_last_two_per_plant))


```

# write and zip

```{r, output}

file_path = '../intermediate/OrthoDB_cleaned.txt'
data.table::fwrite(combo, file_path, sep = '\t')
root_dir = dirname(file_path)
file_name = basename(file_path)
zip::zip(zipfile = paste0(file_path, ".zip"), 
    files = file_name, 
    root = root_dir)
file.remove(file_path)


```


# sessionInfo

```{r,  echo=TRUE, warning=FALSE, message=FALSE}

sessionInfo()

```


