---
title: "09_fruitTrees"
author: "zagor"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    self_contained: yes
    fig_width: 9
    fig_height: 6
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 5
    number_sections: yes
    theme: flatly
    highlight: tango
  pdf_document:
    toc: yes
    toc_depth: '2'
  word_document:
    toc: yes
    toc_depth: '2'
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = c('png', 'pdf'),  # this embeds pdf and crates scrolable blocks if pdf is first
                      fig.path = "../reports/fruitTrees/", # save each figure here
                      # dev = c('png'), 
                      fig.align = 'center', 
                      dpi = 300,
                      fig.height = 6, 
                      fig.width = 9 ,
                      # warning = FALSE, message = FALSE
                      echo = TRUE, message = TRUE, warning = TRUE
                      )
```



```{r,  echo=TRUE, warning=FALSE, message=FALSE}

rm(list = ls(all = TRUE))
gc()


library(magrittr)
library(data.table)
library(knitr)

`%!in%` = Negate(`%in%`)
`%nin%` = Negate(`%in%`)

setwd(dirname(rstudioapi::getActiveDocumentContext()$path))




```

# Ath gmm

* lowercase, protein based IDs!

```{r, Ath-gmm}

# fp = file.path('..', 'input', 'ath-annot')
fp = file.path('..', 'input', 'Mercator')
# fn = 'ath_Araport11_2018-05-25_mapping.txt.gz'
# fn = 'X4.4_Arabidopsis_thaliana.txt'
fn = 'ath_Mercator4v7_results.txt'
gmm = data.table::fread(file.path(fp, fn), header = TRUE, fill = TRUE)
gmm = gmm[gmm$IDENTIFIER != "''", ]

combined = gmm[, .(
  BINCODE = paste(unique(BINCODE), collapse = " | "),
  NAME = paste(unique(NAME), collapse = " | "),
  DESCRIPTION = paste(unique(DESCRIPTION), collapse = " | ")
), by = IDENTIFIER]

combined$IDENTIFIER = sub("\\'", '', sub("\\..*$", "", toupper(combined$IDENTIFIER)))
combined$BINCODE = sub("\\'", '', combined$BINCODE )
combined$NAME = sub("\\'", '', combined$NAME)
combined$DESCRIPTION = sub("\\'", '', combined$DESCRIPTION)
colnames(combined)[2:4] = paste('ath', colnames(combined)[2:4], sep = '_')

ath.gmm = combined


```


# Ath SKM & annotation

note: some duplicated ids in PSS


```{r, Ath-SKM-annot}

fp = file.path('..', 'input', 'ath-annot', 'Phytozome', 'PhytozomeV12', 
               'early_release', 'Athaliana_447_Araport11', 'annotation')
# fn = 'Araport11_GFF3_genes_transposons.current_utf8_attributes_CB.tsv'
fn = 'Athaliana_447_Araport11.geneName.txt'
gn = data.table::fread(file.path(fp, fn), header = FALSE, fill = TRUE)
colnames(gn)[2] = 'athName'
gn$V1 = sub('\\..*', '', gn$V1)
gn = gn[!duplicated(gn), ]


fn = 'Athaliana_447_Araport11.synonym.txt'
sn = data.table::fread(file.path(fp, fn), header = FALSE, fill = TRUE)
sn[, merged_column := apply(.SD, 1, function(x) {
  # Remove NA and empty strings
  x = x[!is.na(x) & x != ""]
  paste(x, collapse = " | ")
}), .SDcols = 2:ncol(sn)]
# Optionally, remove the original columns V2 to V15
sn[, (2:(ncol(sn)-1)) := NULL]
colnames(sn)[2] = 'athSynonims'
sn$V1 = sub('\\..*', '', sn$V1)
sn = sn[!duplicated(sn), ]


fp = file.path('..', 'input', 'SKM_2025-07-08')
fn = 'rxn-nodes-public.tsv'
pss = data.table::fread(file.path(fp, fn), header = TRUE, fill = TRUE)
ind = grep('^name$|^all_pathways|^short_name$', colnames(pss), value = TRUE)
pss = pss[, ..ind]
ind = grep('\\[', pss$name)
pss = pss[ind, ]

pss[, ids_string := stringr::str_extract(name, "(?<=\\[)[^\\]]+(?=\\])")]
pss[, ids_list := strsplit(ids_string, split = ",")]
max_ids = max(lengths(pss$ids_list))
for (i in seq_len(max_ids)) {
  pss[[paste0("id_", i)]] = sapply(pss$ids_list, function(x) ifelse(length(x) >= i, x[i], NA))
}
pss[, c("ids_string", "ids_list") := NULL]

pss_long = melt(
  pss,
  id.vars = c("name", "all_pathways", 'short_name'),       # Columns to keep as is
  measure.vars = patterns("^id_"),           # Columns to melt (all starting with "id_")
  variable.name = "id_num",                   # Name for the melted variable column
  value.name = "id"                           # Name for the melted value column
)

pss_long = pss_long[!is.na(id) & id != ""]
pss_long[, id_num := NULL]
pss_long[, name := NULL]
pss_long$id = sub('\\..*', '', pss_long$id)
pss_long = pss_long[!duplicated(pss_long), ]
table(duplicated(pss_long$id))

pss_long %>%
  dplyr::filter(id %in% id[duplicated(id)] & stringr::str_starts(id, "^AT")) %>%
  dplyr::arrange(id) %>%
  print()


pss_long = pss_long %>%
  dplyr::filter(stringr::str_starts(id, "AT")) %>%
  dplyr::group_by(id) %>%
  dplyr::summarise(
    dplyr::across(
      .cols = dplyr::everything(),
      .fns = ~ {
        vals = unique(na.omit(.))
        if (length(vals) > 1) paste(vals, collapse = " | ")
        else if (length(vals) == 1) vals
        else NA_character_
      }
    ),
    .groups = "drop"
  )



```



# Abbreviations

| Plant Name                    | Label         | JCVI-MCScan | Compara Plants | Plaza | OrthoDB | FastOma | RBH  | Mercator |
|-------------------------------|---------------|-------------|----------------|-------|---------|---------|------|----------|
| Malus domestica               | apple         | *mdo_GDv1*  | *malus_domestica_golden*| mdo   | mdo     | mdo     | mdo  | mdo      |
|                               |               | *mdo_HChap1*|                |       |         |         |      |          |
| Prunus persica                | ppe           | ppe         | *prunus_persica*       | ppe   | ppe     | *pper*  | ppe  |  pper        |
| Prunus dulcis / P. amygdalus  | almond        | *almond*      | *prunus_dulcis*      |       | pdul    | pdul    | pdul | pdul     |
| Prunus avium                  | wild cherry   | *wildcherry*  | *prunus_avium*       |       | pavi    | pavi    | pavi | pavi     |
| Prunus armeniaca              | apricot       | *apricot*    |                       |      | parm    | parm    | parm | parm     |
| Prunus cerasifera             | cherry plum   | *cherryplum*  |                      |     |         | pcer    | pcer | pcer     |
| Pyrus                         | pear          | *pear*        |                      |     |         | pcox    | pcox | pcox     |
| Prunus sibirica               | Siberian apricot | *siberianapricot* |               |     |         | psib    | psib | psib     |




```{r, params}

# in OrthoDB and PLAZA

# mdo_HChap1/mdo/mdo
# ppe/ppe/pper


# in OrthoDB, no PLAZA
# pdul/pdul/pdul
# wildcherry/pavi/pavi
# apricot/parm/parm



# not in OrthoDB or PLAZA
# pear/pcox/pcox
# cherryplum/pcer/pcer
# siberianapricot/psib/psib




```




```{r, mdo, results='asis'}


params_list <- list(

  plantName1 = 'mdo'
  , # change name - PLAZA, OrthoDB, RBH
  plantName2 = 'malus_domestica_golden'
  , # change name - compara # sources
  plantName3 = 'mdo_HChap1'
  ,  # change name - MCScanX # sources
  plantName4 = 'mdo'
  ,  # change name - FastOMA # sources
  
  plantDirIn = "mdo_apple"
  ,# inconsistent-IDs, orthofinder
  plantNameOut = "apple"
  ,
  plantDirOut = file.path('..', 'reports', 'fruitTrees', "apple")
  ,
  
  pattern_in = "\\.[^.]*$"
  , # everythin after the last dot
  pattern_out = ""
  , # all-IDs
  compara_pattern_in1 = '\\..*'
  ,
  compara_pattern_out1 = ""
  ,
  compara_pattern_in2 = ".*_"
  ,
  compara_pattern_out2 = ""
  ,
  plaza_pattern_in1 = '\\..*'
  ,
  plaza_pattern_in2 = ".* "
  ,
  
  
  ref_genome = "g.Honeycrisp_HAP1_braker1+2_combined_fullSupport_longname_filtered.pep"
  , # inconsistent-IDs
  
  mercator = 'mdo_Mercator4v7_results.txt'
  , # plant-gmm
  mercatorPatternIn1 = "[\u2018\u2019\u201C\u201D']"
  , # plant-gmm
  mercatorPatternOut1 = ""
  , # plant-gmm
  mercatorPatternIn2 = "a.g"
  , # plant-gmm
  mercatorPatternOut2 = "A.g"
  , # plant-gmm
  flag1 = 1
  ,
  flag2 = 1
  ,
  flag3 = FALSE

)

env <- new.env()
list2env(params_list, envir = env)

child_content <- knitr::knit_child("08_fruitTrees-child1.rmd", envir = env, quiet = FALSE)
cat(child_content)



```


```{r, ppe, results='asis'}

params_list <- list(
  plantName1 = 'ppe', # change name - PLAZA, OrthoDB, RBH
  plantName2 = 'prunus_persica', # change name - compara # sources
  plantName3 = 'ppe',  # change name - MCScanX # sources
  plantName4 = 'pper',  # change name - FastOMA # sources
  
  plantDirIn = "ppe_peach", # inconsistent-IDs, orthofinder
  plantNameOut = "peach",
  plantDirOut = file.path('..', 'reports', 'fruitTrees', "peach"),

  pattern_in = "\\.[^.]*$", # everythin after the last dot
  pattern_out = "", # all-IDs
  compara_pattern_in1 = "",
  compara_pattern_out1 = "",
  compara_pattern_in2 = "",
  compara_pattern_out2 = "",
  plaza_pattern_in1 = "",
  plaza_pattern_in2 = "",
  
  ref_genome = "PLAZA_proteome.selected_transcript.ppe", # inconsistent-IDs, orthofinder for OrthoDB
  
  mercator = 'pper_Mercator4v7_results.txt', # plant-gmm
  mercatorPatternIn1 = "[\u2018\u2019\u201C\u201D']", # plant-gmm, generic removal of nonsence
  mercatorPatternOut1 = "", # plant-gmm
  mercatorPatternIn2 = "g", # plant-gmm
  mercatorPatternOut2 = "G", # plant-gmm
  flag1 = 1,
  flag2 = 2,
  flag3 = FALSE
)

# note: in compara - geneID and prot ID are completely different

env <- new.env()
list2env(params_list, envir = env)

child_content <- knitr::knit_child("08_fruitTrees-child1.rmd", envir = env, quiet = FALSE)
cat(child_content)


```

```{r, pdul, results='asis'}

params_list <- list(
  
  plantName1 = 'pdul'
  , # change name - PLAZA, OrthoDB, RBH
  plantName2 = 'prunus_dulcis'
  , # change name - compara # sources
  plantName3 = 'almond'
  ,  # change name - MCScanX # sources
  plantName4 = 'pdul'
  ,  # change name - FastOMA # sources
  
  plantDirIn = "pdul_almond"
  , # inconsistent-IDs, orthofinder
  plantNameOut = "almond"
  ,
  plantDirOut = file.path('..', 'reports', 'fruitTrees', "almond")
  ,

  pattern_in = "\\.[^.]*$"
  , # everythin after the last dot
  pattern_out = ""
  , # all-IDs
  compara_pattern_in1 = ""
  ,
  compara_pattern_out1 = ""
  ,
  compara_pattern_in2 = ""
  ,
  compara_pattern_out2 = ""
  ,
  plaza_pattern_in1 = ""
  ,
  plaza_pattern_in2 = ""
  ,
  
  ref_genome = "Texas_F1_protein"
  , # inconsistent-IDs, orthofinder for OrthoDB
  
  mercator = 'pdul_Mercator4v7_results.txt'
  , # plant-gmm
  mercatorPatternIn1 = "[\u2018\u2019\u201C\u201D']"
  , # plant-gmm, generic removal of nonsence
  mercatorPatternOut1 = ""
  , # plant-gmm
  mercatorPatternIn2 = "([fg])"
  , # plant-gmm
  mercatorPatternOut2 = "\\U\\1" # plant-gmm
  ,
  flag1 = 2
  ,
  flag2 = 2
  ,
  flag3 = FALSE
)

# note: in compara - geneID and prot ID are completely different

env <- new.env()
list2env(params_list, envir = env)

child_content <- knitr::knit_child("08_fruitTrees-child1.rmd", envir = env, quiet = FALSE)
cat(child_content)


```


```{r, pavi, results='asis'}

params_list <- list(
  
  plantName1 = 'pavi'
  , # change name - PLAZA, OrthoDB, RBH
  plantName2 = 'prunus_avium'
  , # change name - compara # sources
  plantName3 = 'wildcherry'
  ,  # change name - MCScanX # sources
  plantName4 = 'pavi'
  ,  # change name - FastOMA # sources
  
  plantDirIn = "pavi_wildCherry"
  , # inconsistent-IDs, orthofinder
  plantNameOut = "wildcherry"
  ,
  plantDirOut = file.path('..', 'reports', 'fruitTrees', "wildcherry")
  ,

  pattern_in = "-[^-]*$"
  , # everythin after the last dot
  pattern_out = ""
  , # all-IDs
  compara_pattern_in1 = "^(Pav_(sc|co)\\d+\\.\\d+_g\\d+\\.\\d+\\.(br|mk)).*" # instead of ids has some strange concatenation
  ,
  compara_pattern_out1 = "\\1"
  ,
  compara_pattern_in2 = ''
  ,
  compara_pattern_out2 = ''
  ,
  plaza_pattern_in1 = ""
  ,
  plaza_pattern_in2 = ""
  ,
  
  ref_genome = "Prunus_avium_Tieton.proteins"
  , # inconsistent-IDs, orthofinder for OrthoDB
  
  mercator = 'pavi_Mercator4v7_results.txt'
  , # plant-gmm
  mercatorPatternIn1 = "[\u2018\u2019\u201C\u201D']"
  , # plant-gmm, generic removal of nonsence
  mercatorPatternOut1 = ""
  , # plant-gmm
  mercatorPatternIn2 = "Fun"
  , # plant-gmm
  mercatorPatternOut2 = "FUN" # plant-gmm
  ,
  flag1 = 2
  ,
  flag2 = 1
  ,
  flag3 = TRUE # compara$Ortholog contains mrna space gene 
)

# note: in compara - geneID and prot ID are completely different

env <- new.env()
list2env(params_list, envir = env)

child_content <- knitr::knit_child("08_fruitTrees-child1.rmd", envir = env, quiet = FALSE)
cat(child_content)


```




```{r}

# Step 1: params_list
# params_list <- list(
# ...
# )
# 
# Step 2: YAML header in 09_fruitTrees-child.Rmd
# ---
# title: "fruitTrees Child"
# output: html_document
# params:
#   plantName1: NULL
#   plantName2: NULL
#   plantName3: NULL
#   plantName4: NULL
#   plantDirIn: NULL
#   plantNameOut: NULL
#   plantDirOut: NULL
#   pattern_in: NULL
#   pattern_out: NULL
#   compara_pattern_in1: NULL
#   compara_pattern_in2: NULL
#   plaza_pattern_in1: NULL
#   plaza_pattern_in2: NULL
#   ref_genome: NULL
#   mercator: NULL
#   mercatorPatternIn1: NULL
#   mercatorPatternOut1: NULL
#   mercatorPatternIn2: NULL
#   mercatorPatternOut2: NULL
# ---
# 
# 
# access params in the script like:
# params$plantName1
# params$plantDirOut
# 
# Step 3: Call render() like
# rmarkdown::render(
#   input = "09_fruitTrees-child.Rmd",
#   params = params_list,
#   envir = new.env(parent = globalenv()),  # optional: isolate execution
#   output_format = "html_document",
#   quiet = FALSE
# )
# 
# 
# This will:
# Inject params_list into params$...
# Knit the child Rmd in a separate process
# Print progress to the console (quiet = FALSE)
# Save an .html file to the working directory

```



# SessionInfo

```{r,  echo=TRUE, warning=FALSE, message=FALSE}

sessionInfo()

```


